core/src/main/java/org/apache/iceberg/io/ObjectModelFactory.java

```diff
@@ -0,0 +1,126 @@
+ * while the object model determines the in-memory representation used for the parsed data.
+ * Together, these provide a consistent API for consuming data files while optimizing for specific
+ * processing engines.
+ *
+ * <p>Iceberg provides some built-in object models and processing engines can implement custom
+ * object models to integrate with Iceberg's file reading and writing capabilities.
+ *
+ * @param <E> input schema type used when converting input data to the file format
+ * @param <D> output type used for reading data, and input type for writing data and deletes
+ */
+public interface ObjectModelFactory<E, D> {
+  /** The file format which is read/written by the object model. */
+  FileFormat format();
+
+  /**
+   * Returns the unique identifier for the object model implementation processed by this factory.
+   *
+   * <p>The object model names (for example: "generic", "spark", "spark-vectorized", "flink",
+   * "arrow") act as a contract specifying the expected data structures for both reading (converting
+   * file formats into output objects) and writing (converting input objects into file formats).
+   * This ensures proper integration between Iceberg's storage layer and processing engines.
+   *
+   * @return string identifier for this object model implementation
+   */
+  String objectModelName();
+
+  /**
+   * Creates a writer builder for standard data files.
+   *
+   * <p>The returned {@link WriteBuilder} configures and creates a writer that converts input
+   * objects into the file format supported by this factory for regular data content.
+   *
+   * <p>The builder follows the fluent pattern for configuring writer properties like compression,
+   * encryption, row group size, and other format-specific options.
+   *
+   * @param outputFile destination for the written data
+   * @return configured writer builder for standard data files
+   * @param <B> the concrete builder type for method chaining
+   */
+  <B extends WriteBuilder<B, E, D>> B dataWriteBuilder(OutputFile outputFile);
+
+  /**
+   * Creates a writer builder for equality delete files.
+   *
+   * <p>The returned {@link WriteBuilder} configures and creates a writer that converts input
+   * objects into the file format supported by this factory for equality delete content.
+   *
+   * <p>Equality delete files contain records that identify rows to be deleted based on equality
+   * conditions.
+   *
+   * <p>The builder follows the fluent pattern for configuring writer properties like compression,
+   * encryption, row group size, and other format-specific options.
+   *
+   * @param outputFile destination for the written equality delete data
+   * @return configured writer builder for equality delete files
+   * @param <B> the concrete builder type for method chaining
+   */
+  <B extends WriteBuilder<B, E, D>> B equalityDeleteWriteBuilder(OutputFile outputFile);
```

Comment:
The file formats expose one builder for both delete types. Is the intent here to avoid needing a separate builder type for data, deletes, and equality deletes? (That seems reasonable to me, but I'm trying to understand the choice.)